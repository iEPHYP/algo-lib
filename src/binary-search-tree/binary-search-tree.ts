import { Queue } from '@/queue';

export class BstNode<TValue = number> {
  public value: TValue;
  public leftChild: BstNode<TValue> | null;
  public rightChild: BstNode<TValue> | null;
  public duplicatesCount = 0;

  constructor(value: TValue) {
    this.value = value;
    this.leftChild = null;
    this.rightChild = null;
  }
}

export class BinarySearchTree<TValue = number> {
  public root: BstNode<TValue> | null = null;

  constructor(root?: BstNode<TValue>) {
    this.root = root ?? null;
  }

  public insert(value: TValue): BstNode<TValue> {
    const node = new BstNode(value);
    if (!this.root) {
      this.root = node;

      return node;
    }

    let currentComparingNode = this.root;

    while (true) {
      if (value < currentComparingNode.value) {
        if (currentComparingNode.leftChild) {
          currentComparingNode = currentComparingNode.leftChild;
        } else {
          currentComparingNode.leftChild = node;
          break;
        }
      } else if (value > currentComparingNode.value) {
        if (currentComparingNode.rightChild) {
          currentComparingNode = currentComparingNode.rightChild;
        } else {
          currentComparingNode.rightChild = node;
          break;
        }
      } else {
        // TODO: research the possible realizations for duplicates and realize them
        currentComparingNode.duplicatesCount++;

        return currentComparingNode;
      }
    }

    return node;
  }

  public find(value: TValue): BstNode<TValue> | null {
    if (this.root === null) {
      return null;
    }

    let currentComparingNode = this.root;

    while (true) {
      if (value === currentComparingNode.value) {
        return currentComparingNode;
      } else if (value < currentComparingNode.value) {
        if (currentComparingNode.leftChild) {
          currentComparingNode = currentComparingNode.leftChild;
        } else {
          return null;
        }
      } else if (value > currentComparingNode.value) {
        if (currentComparingNode.rightChild) {
          currentComparingNode = currentComparingNode.rightChild;
        } else {
          return null;
        }
      }
    }
  }

  public has(value: TValue): boolean {
    return !!this.find(value);
  }

  public bfs({ onNodeVisit }: { onNodeVisit: (node: BstNode<TValue>) => void }) {
    if (!this.root) {
      return;
    }

    const queue = new Queue<BstNode<TValue> | null>();
    queue.enqueue(this.root);

    while (!queue.empty()) {
      const currentNode = queue.dequeue();

      if (!currentNode) {
        break;
      }

      onNodeVisit(currentNode);

      [currentNode.leftChild, currentNode.rightChild].forEach((childNode) => {
        childNode && queue.enqueue(childNode);
      });
    }
  }

  // TODO: realize pretty printing
  // public toString() {}
}
