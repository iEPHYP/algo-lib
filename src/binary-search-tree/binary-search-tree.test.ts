import { BinarySearchTree } from './binary-search-tree';

describe('BinarySearchTree', () => {
  let bst = new BinarySearchTree();

  const fillBst = () => {
    bst.insert(10);
    bst.insert(5);
    bst.insert(14);
    bst.insert(3);
    bst.insert(9);
    bst.insert(12);
  };

  beforeEach(() => {
    bst = new BinarySearchTree();
  });

  it('inserts value properly', () => {
    bst.insert(10);
    bst.insert(5);
    bst.insert(14);
    bst.insert(3);
    bst.insert(9);
    bst.insert(12);

    expect(bst).toMatchSnapshot();
  });

  it('inserts value properly for duplicated items', () => {
    bst.insert(10);
    bst.insert(5);
    bst.insert(14);
    bst.insert(14);
    bst.insert(3);
    bst.insert(9);
    bst.insert(12);

    expect(bst).toMatchSnapshot();
  });

  it('finds node by value properly', () => {
    expect(bst.find(4)).toMatchSnapshot('on empty tree');

    bst.insert(10);
    expect(bst.find(4)).toMatchSnapshot('on tree with root only');
    expect(bst.has(4)).toBe(false);
    expect(bst.find(10)).toMatchSnapshot('on tree with root only');
    expect(bst.has(10)).toBe(true);
    bst.insert(5);
    bst.insert(14);
    bst.insert(3);
    bst.insert(9);
    bst.insert(12);

    expect(bst.find(10)).toMatchSnapshot('for 10');
    expect(bst.has(10)).toBe(true);
    expect(bst.find(9)).toMatchSnapshot('for 9');
    expect(bst.has(9)).toBe(true);
    expect(bst.find(12)).toMatchSnapshot('for 12');
    expect(bst.has(12)).toBe(true);
    expect(bst.find(13)).toMatchSnapshot('for 13');
    expect(bst.has(13)).toBe(false);
    expect(bst.find(16)).toMatchSnapshot('for 16');
    expect(bst.has(16)).toBe(false);
  });

  it('bfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.bfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('preorderDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.preorderDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('postorderDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.postorderDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('inorderDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.inorderDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('mrlDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.mrlDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('rmlDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.rmlDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });

  it('rlmDfs works properly', () => {
    fillBst();

    const nodeVisitOrder: number[] = [];
    bst.rlmDfs({
      onNodeVisit: (node) => {
        nodeVisitOrder.push(node.value);
      },
    });

    expect(nodeVisitOrder).toMatchSnapshot();
  });
});
