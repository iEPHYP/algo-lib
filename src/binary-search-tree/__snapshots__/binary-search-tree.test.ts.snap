// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinarySearchTree bfs works properly 1`] = `
[
  10,
  5,
  14,
  3,
  9,
  12,
]
`;

exports[`BinarySearchTree finds node by value properly: for 9 1`] = `
BstNode {
  "duplicatesCount": 0,
  "leftChild": null,
  "rightChild": null,
  "value": 9,
}
`;

exports[`BinarySearchTree finds node by value properly: for 10 1`] = `
BstNode {
  "duplicatesCount": 0,
  "leftChild": BstNode {
    "duplicatesCount": 0,
    "leftChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": null,
      "rightChild": null,
      "value": 3,
    },
    "rightChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": null,
      "rightChild": null,
      "value": 9,
    },
    "value": 5,
  },
  "rightChild": BstNode {
    "duplicatesCount": 0,
    "leftChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": null,
      "rightChild": null,
      "value": 12,
    },
    "rightChild": null,
    "value": 14,
  },
  "value": 10,
}
`;

exports[`BinarySearchTree finds node by value properly: for 12 1`] = `
BstNode {
  "duplicatesCount": 0,
  "leftChild": null,
  "rightChild": null,
  "value": 12,
}
`;

exports[`BinarySearchTree finds node by value properly: for 13 1`] = `null`;

exports[`BinarySearchTree finds node by value properly: for 16 1`] = `null`;

exports[`BinarySearchTree finds node by value properly: on empty tree 1`] = `null`;

exports[`BinarySearchTree finds node by value properly: on tree with root only 1`] = `null`;

exports[`BinarySearchTree finds node by value properly: on tree with root only 2`] = `
BstNode {
  "duplicatesCount": 0,
  "leftChild": null,
  "rightChild": null,
  "value": 10,
}
`;

exports[`BinarySearchTree inserts value properly 1`] = `
BinarySearchTree {
  "root": BstNode {
    "duplicatesCount": 0,
    "leftChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 3,
      },
      "rightChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 9,
      },
      "value": 5,
    },
    "rightChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 12,
      },
      "rightChild": null,
      "value": 14,
    },
    "value": 10,
  },
}
`;

exports[`BinarySearchTree inserts value properly for duplicated items 1`] = `
BinarySearchTree {
  "root": BstNode {
    "duplicatesCount": 0,
    "leftChild": BstNode {
      "duplicatesCount": 0,
      "leftChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 3,
      },
      "rightChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 9,
      },
      "value": 5,
    },
    "rightChild": BstNode {
      "duplicatesCount": 1,
      "leftChild": BstNode {
        "duplicatesCount": 0,
        "leftChild": null,
        "rightChild": null,
        "value": 12,
      },
      "rightChild": null,
      "value": 14,
    },
    "value": 10,
  },
}
`;
